ocaml/xapi-guard/lib/varstored_interface.ml

```diff
@@ -99,53 +99,81 @@ let serve_forever_lwt_callback rpc_fn path _ req body =
       in
       Cohttp_lwt_unix.Server.respond_string ~status:`Method_not_allowed ~body ()
 
-let serve_forever_lwt_callback_vtpm backend _path _ req body =
+module StringMap = Map.Make(String)
+
+(* this is not efficient if we have lots of keys,
+   but using a map on the XAPI API would just move the inefficiency to the XAPI DB instead
+*)
+
+let deserialize t =
+    match t |> Yojson.Safe.from_string with
+    | `Assoc alist ->
+        StringMap.of_seq
+          (alist
+          |> List.to_seq
+          |> Seq.map @@ fun (k, v) ->
+             ( k |> Base64.decode_exn
+             , v
+               |> Yojson.Safe.Util.to_string
+               |> Base64.decode_exn
+             )
+          )
+    | _ ->
+        invalid_arg "malformed JSON"
+
+let serialize t =
+    let alist =
+        List.of_seq
+          (t
+          |> StringMap.to_seq
+          |> Seq.map @@ fun (k, v) ->
+             (* we pass it through XAPI APIs, had to encode *)
+             ( k |> Base64.encode_string
+             , `String (v |> Base64.encode_string)
+             )
+          )
+      in
+       `Assoc alist |> Yojson.Safe.to_string
+
+let serve_forever_lwt_callback_vtpm ~cache mutex vtpm _path _ req body =
   let uri = Cohttp.Request.uri req in
-  let module VB = VTPM_backend in
+  (* in case the connection is interrupted/etc. we may still have pending operations,
+     so use a per vTPM mutex to ensure we really only have 1 pending operation at a time for a vTPM
+  *)
+  Lwt_mutex.with_lock mutex @@ fun () ->
   (* TODO: some logging *)
   match (Cohttp.Request.meth req, Uri.path uri) with
-  | `GET, "/" ->
-      (* swtpm doesn't have to use this, it is added for completeness/debugging purposes *)
-      let* keys = VB.list backend in
-      let body =
-        `Tuple (keys |> List.map @@ fun k -> `String (VB.Key.to_string k))
-        |> Yojson.Safe.to_string
-      in
-      let headers =
-        Cohttp.Header.of_list [("Content-Type", "application/json")]
-      in
-      Cohttp_lwt_unix.Server.respond_string ~headers ~status:`OK ~body ()
   | `GET, key -> (
-      let* value_opt = VB.get backend (VB.Key.of_string_exn key) in
-      match value_opt with
+      let* contents = with_xapi ~cache @@ VTPM.get_contents ~self:vtpm in
+      let map = deserialize contents in
```

Comment:
I didn't expect the contents to be in json format, is there any explanation on the format of the values stored in the database?

Reply:
SWTPM can store multiple key-value pairs for a single vtpm. I think in practice that key is only a single thing with a binary value, but I'll have to check.
We should indeed use something more efficient than json here.